numpyro/contrib/distributions/discrete.py

```diff
@@ -20,23 +31,125 @@ def _to_logits_bernoulli(probs):
 
+    @lazy_property
+    def probs(self):
+        return _to_probs_bernoulli(self.logits)
+
+    @property
+    def mean(self):
+        return np.broadcast_to(self.probs, self.batch_shape)
+
+    @property
+    def variance(self):
+        return np.broadcast_to(self.probs * (1 - self.probs), self.batch_shape)
+
+
+class Binomial(Distribution):
+    arg_constraints = {'total_count': constraints.nonnegative_integer,
+                       'probs': constraints.unit_interval}
+
+    def __init__(self, probs, total_count=1, validate_args=None):
+        self.probs, self.total_count = promote_shapes(probs, total_count)
+        super(Binomial, self).__init__(batch_shape=np.shape(self.probs), validate_args=validate_args)
+
+    def sample(self, key, size=()):
+        return binomial(key, self.probs, n=self.total_count, shape=size + self.batch_shape)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        dtype = get_dtypes(self.probs)[0]
+        value = lax.convert_element_type(value, dtype)
+        total_count = lax.convert_element_type(self.total_count, dtype)
+        log_factorial_n = gammaln(total_count + 1)
+        log_factorial_k = gammaln(value + 1)
+        log_factorial_nmk = gammaln(total_count - value + 1)
+        return (log_factorial_n - log_factorial_k - log_factorial_nmk +
+                xlogy(value, self.probs) + xlog1py(total_count - value, -self.probs))
+
+    @property
+    def mean(self):
+        return np.broadcast_to(self.total_count * self.probs, self.batch_shape)
+
+    @property
+    def variance(self):
+        return np.broadcast_to(self.total_count * self.probs * (1 - self.probs), self.batch_shape)
+
+
+class BinomialWithLogits(Distribution):
+    arg_constraints = {'total_count': constraints.nonnegative_integer,
+                       'logits': constraints.real}
+
+    def __init__(self, logits, total_count=1, validate_args=None):
+        self.logits, self.total_count = promote_shapes(logits, total_count)
+        super(BinomialWithLogits, self).__init__(batch_shape=np.shape(self.probs), validate_args=validate_args)
+
+    def sample(self, key, size=()):
+        return binomial(key, self.probs, n=self.total_count, shape=size + self.batch_shape)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        dtype = get_dtypes(self.logits)[0]
+        value = lax.convert_element_type(value, dtype)
+        total_count = lax.convert_element_type(self.total_count, dtype)
+        log_factorial_n = gammaln(total_count + 1)
+        log_factorial_k = gammaln(value + 1)
+        log_factorial_nmk = gammaln(total_count - value + 1)
+        # Note that: torch.log1p(-self.probs)) = - torch.log1p(self.logits.exp()))
+        return (log_factorial_n - log_factorial_k - log_factorial_nmk +
+                value * self.logits - total_count * np.log1p(np.exp(self.logits)))
```

Comment:
It might be better to address the overflow problem when logits is large. You can see [my implementation here](https://github.com/pyro-ppl/numpyro/pull/101/files#diff-fe7dcb1e0f5ffa51f99755b000e28ccbR87). But I am not sure if it resolves the discontinuity problem which we have in PyTorch previously.

Reply:
Makes sense, let me play around with your solution and put it in.